server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -382,16 +384,37 @@ synchronized void stopWatcher() {
         }
     }
 
-    private WatchKey enableSettingsWatcher(WatchKey previousKey, Path settingsDir) throws IOException {
+    // package private for testing
+    long retryDelayMillis(int failedCount) {
+        assert failedCount < 31;
+        return 10 * (1 << failedCount) + Randomness.get().nextInt(10);
+    }
+
+    // package private for testing
+    WatchKey enableSettingsWatcher(WatchKey previousKey, Path settingsDir) throws IOException, InterruptedException {
         if (previousKey != null) {
             previousKey.cancel();
         }
-        return settingsDir.register(
-            watchService,
-            StandardWatchEventKinds.ENTRY_MODIFY,
-            StandardWatchEventKinds.ENTRY_CREATE,
-            StandardWatchEventKinds.ENTRY_DELETE
-        );
+        WatchKey watchKey = null;
+
+        for (int i = 0; i < REGISTER_RETRY_COUNT; i++) {
+            try {
+                watchKey = settingsDir.register(
+                    watchService,
+                    StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.ENTRY_CREATE,
+                    StandardWatchEventKinds.ENTRY_DELETE
+                );
+                break;
+            } catch (IOException e) {
+                if (i == REGISTER_RETRY_COUNT - 1) {
```

Comment:
Why the `-1`?

Reply:
i = 0, 1, 2, 3, 4, that's 5 attempts.

Reply:
Ah yes, I was puzzled about the interaction with the loop exit condition (which would set i to 5 on exiting)